Add Quarkus Vault Authentication Guide

vsevel opened the pull request:
This provides a dedicated guide that covers all currently supported Vault Authentication methods:
 - direct token (including variant with response wrapping)
 - userpass
 - approle (including variant with response wrapping)
 - kubernetes

Fixes #4846

vsevel committed:
Quarkus Vault Authentication Guide - Fixes #4846

lordofthejars reviewed (approved):
Really nice document that will allow us to add more strategies in the future.

lordofthejars commented:
@vsevel one comment shouldn't this PR contains removing the lines of the previous docs?

sberyozkin commented:
@lordofthejars 

> one comment shouldn't this PR contains removing the lines of the previous docs?

Can you clarify please ? Thanks

lordofthejars commented:
@sberyozkin I mean that some of the content added in this doc is duplicated in other Vault guides, so maybe they should be removed as well, and this PR could be used for this.

sberyozkin commented:
@lordofthejars Good point, I suppose the those docs can provide some basic relevant info with the link to the main Vault authentication doc

vsevel commented:
@lordofthejars @sberyozkin thanks for your encouraging comments :-)

> Really nice document that will allow us to add more strategies in the future.

yes definitely. for instance for the cert auth.

> I mean that some of the content added in this doc is duplicated in other Vault guides, so 
maybe they should be removed as well, and this PR could be used for this.

actually there is almost no overlap, because in the base guide I activate the `userpass` auth method and set up the user with a note specifying that other auth methods exist but without any details (client token, approle and k8s); and in the vault-auth guide I say in the `userpass` section that I expect the `userpass` method to be already enabled from the base guide. if we consider the quarkus documentation to be guides (and not reference doc), then I think it is fine.

also in the vault-auth guide I set the base guide to be a prerequisite (allowing to reuse the configured vault + the app). because I think the base guide is what people will want to discover first. and we need to provide a minimal vault auth setup (`userpass` in this case). the base guide needs to be standalone.

one thing we could do in the base guide is to remove the _The Vault extension also supports alternate authentication methods such as:_ note and replace it with: _Other auth methods are supported. Check the vault auth guide for details_.

what yo you think?

gsmet commented on docs/src/main/asciidoc/vault-auth.adoc:
```diff
@@ -0,0 +1,433 @@
+Working with Vault is typically a 2 step process:
+
+* Logging in, which returns a client token
+* Start using Vault using the client token, within the limits of what is allowed in policies associated with the token
```
```suggestion
* Start using Vault using the client token, within the limits of what is allowed by the policies associated with the token
```
?

gsmet commented on docs/src/main/asciidoc/vault-auth.adoc:
```diff
@@ -0,0 +1,433 @@
+* JDK 1.8+ installed with `JAVA_HOME` configured appropriately
+* Apache Maven {maven-version}
+* Docker installed
+* For the Kubernetes authentication: A Kubernetes distribution to deploy the Quarkus application (minishift, k3s, docker desktop, …​)
```
```suggestion
* For the Kubernetes authentication: A Kubernetes distribution to deploy the Quarkus application (Minishift, K3s, Docker Desktop, …​)
```

gsmet commented on docs/src/main/asciidoc/vault-auth.adoc:
```diff
@@ -0,0 +1,433 @@
+
+== Token Authentication
+
+We assume there is a Vault container and a PostgresQL container running from the {base-guide}, and the root token is known.
```
```suggestion
We assume there is a Vault container and a PostgreSQL container running from the {base-guide}, and the root token is known.
```

gsmet commented on docs/src/main/asciidoc/vault-auth.adoc:
```diff
@@ -0,0 +1,433 @@
+/ # export VAULT_TOKEN={root-token}
+----
+
+Create a token for policy `vault-quickstart-policy`:
```
```suggestion
Create a token for the `vault-quickstart-policy` policy:
```

gsmet commented on docs/src/main/asciidoc/vault-auth.adoc:
```diff
@@ -0,0 +1,433 @@
+
+=== Client Token using Response Wrapping
+
+One drawback of this approach is that you expose a secret piece of information (.i.e the token) that can give access
```
```suggestion
One drawback of this approach is that you expose a secret piece of information (i.e. the token) that can give access
```

gsmet commented on docs/src/main/asciidoc/vault-auth.adoc:
```diff
@@ -0,0 +1,433 @@
+If the _Wrapping Token_ gets stolen and unwrapped, we will notice immediately because the legitimate application
+will get an error saying that the token is invalid.
+
+With that in mind, let’s create a new token and wrap it inside a _Wrapping Token_ with a ttl of 1 minute:
```
```suggestion
With that in mind, let’s create a new token and wrap it inside a _Wrapping Token_ with a TTL of 1 minute:
```

gsmet commented on docs/src/main/asciidoc/vault-auth.adoc:
```diff
@@ -0,0 +1,433 @@
+
+Compile and run the application *without the tests*, you should be able now to curl the private key `{private-key}` as before.
+
+Stop the application, and execute tests with `./mvnw test`. They should fail with error:
```
```suggestion
Stop the application, and execute tests with `./mvnw test`. They should fail with the following error:
```

gsmet commented on docs/src/main/asciidoc/vault-auth.adoc:
```diff
@@ -0,0 +1,433 @@
+[NOTE]
+====
+Userpass supports response wrapping as well for the `password` property, although it is more
+unsual to use this approach as response wrapping typically involves a technical workflow,
```
```suggestion
unusual to use this approach as response wrapping typically involves a technical workflow,
```
or usual? but I would say the former.

gsmet commented on docs/src/main/asciidoc/vault-auth.adoc:
```diff
@@ -0,0 +1,433 @@
+== Kubernetes Authentication
+
+Vault provides an integration with Kubernetes to allow containers to authenticate with Vault using their
+Kubernetes jwt token located in `/var/run/secrets/kubernetes.io/serviceaccount`.
```
```suggestion
Kubernetes JWT token located in `/var/run/secrets/kubernetes.io/serviceaccount`.
```

gsmet commented on docs/src/main/asciidoc/vault-auth.adoc:
```diff
@@ -0,0 +1,433 @@
+Kubernetes jwt token located in `/var/run/secrets/kubernetes.io/serviceaccount`.
+
+The setup is more involved than with the other auth methods because we need to allow Vault to talk
+to the master api to be able to validate the jwt token that the application will authenticate with.
```
```suggestion
to the master API to be able to validate the JWT token that the application will authenticate with.
```

gsmet commented on docs/src/main/asciidoc/vault-auth.adoc:
```diff
@@ -0,0 +1,433 @@
+=== auth-delegator
+
+The first step involves creating a `vault-auth-sa` service account with `auth-delegator`
+role that Vault will use to communicate with the master api.
```
```suggestion
role that Vault will use to communicate with the master API.
```

gsmet commented on docs/src/main/asciidoc/vault-auth.adoc:
```diff
@@ -0,0 +1,433 @@
+
+And apply it: `kubectl apply -f vault-auth-k8s.yml`.
+
+Once the objects are created, we need to capture the jwt token of this service account,
```
```suggestion
Once the objects are created, we need to capture the JWT token of this service account,
```

gsmet commented on docs/src/main/asciidoc/vault-auth.adoc:
```diff
@@ -0,0 +1,433 @@
+=== Deploy the application
+
+Add the following properties to the application (and remove any other authentication Vault property,
+plus replace `<host>` by the ip or name of the local host that is running the Vault and PostgresQL containers,
```
```suggestion
plus replace `<host>` by the ip or name of the local host that is running the Vault and PostgreSQL containers,
```

gsmet reviewed (changes requested):
Awesome! I added some very minor comments.

gsmet commented:
@vsevel should I backport this document to 1.4? Are all the elements here in CR1? Do I need to backport more things? Thanks!

vsevel commented:
dear all 

@lordofthejars @sberyozkin I modified the vault guide to refer to the vault auth guide in the following note: _Quarkus supports additional Vault Authentication methods. Check the {vault-auth-guide} for details._

I also added this precision in the `userpass` section of the vault auth guide: _Normally the userpass auth method should already be enabled from the {base-guide}. If not, execute the following commands now: ..._

> and also link to this doc from the main index page

are you talking about `index.adoc`? I thought this file was obsolete. Is there a place where I should list all vault guides?

@gsmet I have done all the suggested modifications. I have pushed and squashed so if everybody is fine, you can go ahead and merge. you can backport it too, since all the code is already in 1.4 (the last feature was _response wrapping_).

thanks for your comments.

gsmet commented:
Thanks, great addition!
